planner_cspace/test/src/test_start_pose_predictor.cpp

```diff
@@ -0,0 +1,275 @@
+        {
+            StartPosePredictor::Astar::Vec(5, 5, 0),  // Start is (0.0, 0.0, 0.0)
+            StartPosePredictor::Astar::Vec(6, 5, 0),
+            StartPosePredictor::Astar::Vec(7, 5, 0),
+            StartPosePredictor::Astar::Vec(8, 6, 1),   // Curv from (2.0, 0.0, 0.0) to (3.0, 1.0, pi/2)
+            StartPosePredictor::Astar::Vec(9, 5, 2),   // Curv from (3.0, 1.0, 0.0) to (4.0, 0.0, pi)
+            StartPosePredictor::Astar::Vec(10, 5, 2),  // goal is (5.0, 0.0, pi)
+        };
+    paths_["switch_back"] = std::make_pair(path_grid_switch_back, getMetricPath(path_grid_switch_back));
+  }
+
+  nav_msgs::Path getMetricPath(const std::list<GridAstar<3, 2>::Vec>& path_grid) const
+  {
+    const auto path_interpolated = interpolator_.interpolate(path_grid, 0.1, 10);
+
+    nav_msgs::Path result_path;
+    result_path.header.frame_id = "map";
+    for (const auto& pose_grid : path_interpolated)
+    {
+      geometry_msgs::PoseStamped pose;
+      pose.header.frame_id = "map";
+      float x, y, yaw;
+      grid_metric_converter::grid2Metric(map_info_, pose_grid[0], pose_grid[1], pose_grid[2], x, y, yaw);
+      pose.pose.position.x = x;
+      pose.pose.position.y = y;
+      pose.pose.orientation.z = sin(yaw / 2);
+      pose.pose.orientation.w = cos(yaw / 2);
+      result_path.poses.push_back(pose);
+    }
+    return result_path;
+  }
+
+  geometry_msgs::Pose getPose(const double x, const double y, const double yaw) const
+  {
+    geometry_msgs::Pose pose;
+    pose.position.x = x;
+    pose.position.y = y;
+    pose.orientation.z = sin(yaw / 2);
+    pose.orientation.w = cos(yaw / 2);
+    return pose;
+  }
+
+  StartPosePredictor predictor_;
+  StartPosePredictor::Config config_;
+  GridAstar<3, 2>::Gridmap<char, 0x40> cm_;
+  costmap_cspace_msgs::MapMetaData3D map_info_;
+  PathInterpolator interpolator_;
+  std::unordered_map<std::string, std::pair<std::list<StartPosePredictor::Astar::Vec>, nav_msgs::Path>> paths_;
+};
+
+TEST_F(StartPosePredictorTester, EdgeCases)
+{
+  StartPosePredictor::Astar::Vec start_grid;
+  // Empty path
+  EXPECT_FALSE(predictor_.process(getPose(0.0, 0.0, 0.0), cm_, map_info_, nav_msgs::Path(), start_grid));
+
+  // Far from path
+  EXPECT_FALSE(predictor_.process(getPose(0.0, 0.5, 0.0), cm_, map_info_,
+                                  paths_["straight"].second, start_grid));
+
+  // Only the last pose is remaining
+  nav_msgs::Path rotating_path;
+  rotating_path.header.frame_id = "map";
+  rotating_path.poses.resize(1);
+  rotating_path.poses[0].header.frame_id = "map";
+  rotating_path.poses[0].pose.position.x = 0.01;
+  rotating_path.poses[0].pose.position.y = 1.01;
+  rotating_path.poses[0].pose.orientation.z = sin(M_PI / 4);
+  rotating_path.poses[0].pose.orientation.w = cos(M_PI / 4);
+
+  EXPECT_TRUE(predictor_.process(getPose(0.0, 1.0, 1.3), cm_, map_info_, rotating_path, start_grid));
+  EXPECT_EQ(start_grid[0], 5);
+  EXPECT_EQ(start_grid[1], 6);
+  EXPECT_EQ(start_grid[2], 1);
```

Comment:
As there's `==` operator, I think it can be

```suggestion
  EXPECT_EQ(StartPosePredictor::Astar::Vec(5, 6, 1), start_grid);
```

https://github.com/at-wat/neonavigation/blob/fdfea8e0a2eab118622d4eda598d42411804f3c1/planner_cspace/include/planner_cspace/cyclic_vec.h#L163-L170

Reply:
Done